File: src/types/interview.ts
export interface InterviewSession {
  _id: string;
  candidate_id: string;
  start_time: string; // ISO date string
  end_time?: string; // ISO date string
  status: 'initialized' | 'in_progress' | 'completed';
  role: string;
  difficulty: string;
  duration_minutes: number;
  total_score?: number;
  feedback_summary?: string;
  interview_type: string;
  questions_asked: { question_id: string; response_id: string }[];
}

export interface InterviewQuestion {
  _id: string;
  text: string;
  topic: string;
  difficulty: string;
  expected_keywords?: string[];
  ideal_answer?: string;
}

export interface InterviewResponse {
  _id: string;
  session_id: string;
  question_id: string;
  candidate_id: string;
  response_text: string;
  audio_file_path?: string;
  ai_feedback?: string;
  score?: number;
  timestamp: string; // ISO date string
}

export interface UploadedDocument {
  file_name: string;
  file_path: string;
  upload_date: string; // ISO date string
  document_type: 'resume' | 'job_description';
}

File: server/app/models/interview.py
from pydantic import BaseModel, Field
from typing import List, Optional, Any
from datetime import datetime
from bson import ObjectId

class PyObjectId(ObjectId):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v, values):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid objectid")
        return ObjectId(v)

    @classmethod
    def __get_pydantic_json_schema__(cls, field_schema):
        field_schema.update(type="string")


class UploadedDocumentModel(BaseModel):
    file_name: str
    file_path: str
    upload_date: datetime = Field(default_factory=datetime.utcnow)
    document_type: str  # 'resume' or 'job_description'

class UserModel(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    email: str
    name: str
    password_hash: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    roles: List[str] = []
    uploaded_documents: List[UploadedDocumentModel] = []

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
        populate_by_name = True


class InterviewQuestionModel(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    text: str
    topic: str
    difficulty: str
    expected_keywords: Optional[List[str]] = []
    ideal_answer: Optional[str] = None
    embedding: Optional[List[float]] = []
    created_by: Optional[PyObjectId] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
        populate_by_name = True


class InterviewResponseModel(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    session_id: PyObjectId
    question_id: PyObjectId
    candidate_id: PyObjectId
    response_text: str
    audio_file_path: Optional[str] = None
    response_embedding: Optional[List[float]] = []
    ai_feedback: Optional[str] = None
    score: Optional[float] = None
    follow_up_questions: List[Any] = [] # Define a more specific model later if needed
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    duplicacy_score: Optional[float] = None
    similarity_to_ideal: Optional[float] = None

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
        populate_by_name = True


class InterviewSessionModel(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    candidate_id: PyObjectId
    start_time: datetime = Field(default_factory=datetime.utcnow)
    end_time: Optional[datetime] = None
    status: str = "initialized"  # e.g., initialized, in_progress, completed
    role: str
    difficulty: str
    duration_minutes: int
    total_score: Optional[float] = None
    feedback_summary: Optional[str] = None
    interview_type: str = "AI_Agent"
    questions_asked: List[Any] = [] # Contains dicts with question_id and response_id

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
        populate_by_name = True
